### DISCUSSION TITLE
Plastic strain rate coupled with concentration

### DISCUSSION THREAD

**Discussion by sidharthsarmah** [idaholab/moose#22957] [Q&A General] [answered]
Hi @hugary1995 ,

I am trying to implement the last term of this equation and can be reduce as follows (with CL and epsilon_p as the two variables):
<img width="686" alt="image" src="https://user-images.githubusercontent.com/110927986/207381474-0f912c31-7dc5-4abc-b7cd-fe2d7cc94533.png">
<img width="736" alt="image" src="https://user-images.githubusercontent.com/110927986/207379164-4f185e33-d1fa-490f-ad99-970f461a8d46.png">
It consist of a strain rate coupled with concentration_lattice (CL) and the stress-strain variation is as follows:
<img width="638" alt="image" src="https://user-images.githubusercontent.com/110927986/207381277-7ddd98ca-cc23-4be2-a9de-d16370f7b88a.png">
Here is what I have tried:

1. Introduce the effective plastic strain as an ADProperty
```
[Materials]

```
2. Insert the property as an non-linear variable strain_p
```

```
3. Used ADHeatConductionTimeDerivative kernels to solve:
```


[]
```

Could you check if I am correct till here, implementing the kernel, if yes, how do I implemeny the stress-strain relationships?
<img width="638" alt="image" src="https://user-images.githubusercontent.com/110927986/207381277-7ddd98ca-cc23-4be2-a9de-d16370f7b88a.png">

**Comment by hugary1995** [comment, marked as answer]
Take a look at the plasticity/creep regression tests in modules/tensor_mechanics/test/tests. If you set things up correctly, epsilon_p will be available as equivalent_plastic_strain or equivalent_inelastic_strain.

You don't need to project the epsilon_p onto a field, because you don't need its gradients. I suggest you implement your own kernel representing this term. If you don't know how to write a custom kernel, check out the tutorials and the workshop.

**Reply by sidharthsarmah** [reply]
Hi @hugary1995,

Yes, I will try to build a custom kernel, could you suggest, how do I implement any stress-strain relationship as such?

![](https://user-images.githubusercontent.com/110927986/207381277-7ddd98ca-cc23-4be2-a9de-d16370f7b88a.png)

**Reply by hugary1995** [reply]
By "implement", do you mean that you want to write source code to perform the constitutive update, or do you want to use already implemented materials in the tensor_mechanics module? The former is usually what I mean by "implementing" something, but I realize you may be referring to the latter.

Either way, I recommend you to start with the tensor_mechanics documentation, as always. https://mooseframework.inl.gov/modules/tensor_mechanics/index.html
On the right side there is a column listing "advanced features", where you will find descriptions on "Multiple inelastic stress", "Crystal plasticity", "Viscoplasticity", etc.. Then, look at the input files inside `modules/tensor_mechanics/test/tests` for additional examples on how to set up these models.

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , I want to couple the plastic strain `eP` with conc, `cL`. Which is the correct way to do so, the `cL` here is the primary variable?
<img width="216" alt="image" src="https://github.com/idaholab/moose/assets/110927986/611a0bce-fad0-4f51-8a17-5f0775af203b">
```
[Variables]
  [./cL]
  [../]

  [./eP]
  [../]
[]
```

1. Using `ADHeatConductionTimeDerivative`:
```
[Kernels]
  [./CL_time_term]
    type = ADHeatConductionTimeDerivative
    variable = 'cL'
    specific_heat = '1'
    density_name = 'func1'
  [../]

  [./strain_effective]
    type = ADMaterialPropertyValue
    prop_name = 'effective_plastic_strain'
    variable = 'eP'
  [../]


[Materials]
  [./diff_time]
    type = ADParsedMaterial
    expression = 'if(eP < 0.5, (2/3)*cL + eP, (1/2)*cL)'
    args = 'cL	eP'
    property_name = 'func1'
  [../]
```
2. Using `ADCoupledMatTimeDerivative`
```
  [./strainrate_term]
    type = ADCoupledMatTimeDerivative
    variable = 'eP'
    v = 'cL'
    coef = 'func2'
  [../]

  [./strain_effective]
    type = ADMaterialPropertyValue
    prop_name = 'effective_plastic_strain'
    variable = 'eP'
  [../]


[Materials]
  [./diff_time]
    type = ADParsedMaterial
    expression = 'if(eP < 0.5, (2/3)*cL + eP, (1/2)*cL)'
    args = 'cL	eP'
    property_name = 'func2'
  [../]
```
